modules/services/podman-linux/podman-lib.nix

```diff
@@ -1,71 +1,145 @@
-{ lib, ... }:
+{ lib, config, ... }:
 
 with lib;
 
 let
-  primitive = with types; nullOr (oneOf [ bool int str path ]);
-  primitiveAttrs = with types; attrsOf (either primitive (listOf primitive));
+  normalizeKeyValue = k: v:
+    let
+      v' = (if builtins.isBool v then
+        (if v then "true" else "false")
+      else if builtins.isAttrs v then
+        (concatStringsSep ''
 
-  formatPrimitiveValue = value:
-    if isBool value then
-      (if value then "true" else "false")
-    else if isList value then
-      concatStringsSep " " (map toString value)
-    else
-      toString value;
-in {
-  inherit primitive;
-  inherit primitiveAttrs;
-  inherit formatPrimitiveValue;
+          ${k}='' (mapAttrsToList normalizeKeyValue v))
+      else
+        builtins.toString v);
+    in if builtins.isNull v then "" else "${k}=${v'}";
 
-  serviceConfigTypeRules = {
-    Restart = types.enum [ "no" "always" "on-failure" "unless-stopped" ];
-    TimeoutStopSec = types.int;
-  };
-  serviceConfigDefaults = {
-    Restart = "always";
-    TimeoutStopSec = 30;
-    ExecStartPre = null;
+  primitiveAttrs = with types; attrsOf (either primitive (listOf primitive));
+  primitiveList = with types; listOf primitive;
+  primitive = with types; nullOr (oneOf [ bool int str path ]);
+
+  toQuadletIni = generators.toINI {
+    listsAsDuplicateKeys = true;
+    mkKeyValue = normalizeKeyValue;
   };
-  serviceConfigType = with types; attrsOf (either primitive (listOf primitive));
 
-  unitConfigTypeRules = { After = with types; nullOr (listOf str); };
-  unitConfigDefaults = { After = null; };
-  unitConfigType = with types; attrsOf (either primitive (listOf primitive));
+  # meant for ini. favours b when two values are unmergeable
+  deepMerge = a: b:
```

Comment:
this should be achievable in equivalent but less code with a `lib.zipAttrsWith (impl)`. Do you think its worth writing it as that @bamhm182?

Reply:
Unless I'm missing it, I don't think the code difference is that large. I think it's fine the way you have it.

Reply:
Yeah. was playing around and putting the result here for posterity. It would probably look something like this:

```nix
  mergeAttrs = with builtins; name: values:
    let
      realValues = filter (v: v != null) values;
      allAreAttrs = all (v: typeOf v == "set") realValues;
      anyIsList = any isList realValues;
    in
      if length realValues == 0 then
        null
      else if allAreAttrs then
        lib.attrsets.zipAttrsWith mergeAttrs realValues
      else if length realValues == 1 then
        head realValues
      else if anyIsList then
        let
          mergedList = concatLists (map (v:
            if isList v then v else [v]
          ) realValues);
        in
          if length mergedList == 0 then null else mergedList
      else
        last realValues;

  deepMerge = a: b: zipAttrsWith mergeAttrs [a b];
```